### ISSUE TITLE
RStudio setwd to mapped network drive not working Win10

### ISSUE THREAD

**Issue by andzandz11** [rstudio/rstudio#4030]
issue has been reported in the past but without solution: https://support.rstudio.com/hc/en-us/community/posts/200652973-Network-Drives

has been reported here without solution:
https://stackoverflow.com/questions/25268582/rstudio-desktop-0-98-1006-for-windows-unable-to-access-mapped-network-drives

Version 1.1.453 – © 2009-2018 RStudio, Inc.
Mozilla/5.0 (Windows NT 6.2; WOW64) AppleWebKit/538.1 (KHTML, like Gecko) rstudio Safari/538.1 Qt/5.4.1

> setwd("Y:/")
Error in setwd("Y:/") : cannot change working directory

output from choose.dir()
![grafik](https://user-images.githubusercontent.com/19622117/49866591-7a4f2200-fe08-11e8-845d-7a6e24de5f0a.png)

same computer:
![grafik](https://user-images.githubusercontent.com/19622117/49866279-9a321600-fe07-11e8-9a42-7251de986764.png)

output from choose.dir()
![grafik](https://user-images.githubusercontent.com/19622117/49866658-a5d20c80-fe08-11e8-81b2-8640c99e3946.png)

![grafik](https://user-images.githubusercontent.com/19622117/49866343-c64d9700-fe07-11e8-9534-acb1672b6d98.png)

**Comment by ronblum** [comment]
@andzandz11 Thank you for raising this issue. We will review it as we work on development of RStudio. Would you be willing to try out [the preview of the upcoming release](https://www.rstudio.com/products/rstudio/download/preview/) to see if the problem is resolved?

**Comment by JoFAM** [comment]
Some more info, as we have the same problem on our university network. I've noticed a striking difference in the environment variables used by RStudio compared to R itself. The culprit is the HOME environment variable as reported by `Sys.getenv()`. This is given as

//files/loginname/home

in RStudio and 

\\files\loginname\home

in R itself.

Our workaround is to use save as, go to "This PC" and expand, go the the actual drive (H: for example) and find the folder from there. That works. Using save as, then use the shortcut "Home" in the explorer and then browsing to the correct subfolder doesn't. 

This is using the latest version of RStudio (1.2.5001) and R 3.6.1

**Comment by cbrnr** [comment]
I have the same (or at least related) issue on my university network. The startup working directory is set to be the user's home drive mapped to Z:. The output of `getwd()` is `"\\\\some.server.address/fs/home/user"`. Saving a script to that location does not work as it results in a location not found error. My workaround is to manually change to the mapped drive first with `setwd("Z:")`, which leads to a `getwd()` value of `"Z:/"`. Only then does saving a script work correctly. Maybe RStudio should try to use mapped drive letters instead of the underlying network paths?

**Comment by rvino** [comment]
Similar bug. My only solution has been after `choose.dir()` I right-click on 'This PC' and choose the option 'Map to Network Drive' and I map to the network drive. It still doesn't appear, but if I then close the pop-up box and reopen it with `choose.dir()` the network drive is there and is maintained after restarting Rstudio. I've only run into issues after updating the remote server I'm working on and have to do the above process again.

**Comment by kevinushey** [comment]
I believe these issues are resolved in the [preview release](https://www.rstudio.com/products/rstudio/download/preview/) of RStudio (v1.3). Would you be able to download that and give it a shot?

**Comment by rvino** [comment]
Upgraded to version 1.3.944 (R version 3.6.1) and rebooted remote. Same issue unfortunately.

**Comment by gwd666** [comment]
@kevinushey can u pls check my new/additional comments back at https://github.com/rstudio/rstudio/pull/6598

**Comment by kevinushey** [comment]
I think @gtritchie commented in the other post -- sorry, I was incorrect that this is fixed in the current v1.3 preview release; our goal is to bring the fix as part of a later patch release. Sorry for the trouble.

**Comment by lwatanabe** [comment]
I found a way to trick R and RStudio into finding my network drive. Back story...the network drive was accessible through Windows Explorer but R could not see it. Even choose.dir() did not work. When I tried to map the network drive, it was already assigned.

My solution...save a script on the network drive and open the script directly (double-click). It opened with no problem. Now, RStudio can find the drive, even after restarting my machine.

**Comment by gwd666** [comment]
@Iwatanabe you basically are describing a variation of @cbrnr or @JoFAM comments and solutions for mapped networks on WIN
PS: R does not show issues with these drives it was only RStudio Versions 1.2.5033 to 1.3.959; 
the 1.3. Patch versions so far seems to have solved that

**Comment by lwatanabe** [comment]
I did try it in my R installation directly but choose.dir() did not show the drive. Even though I just installed them, I will check that the versions are the most recent with the appropriate patches. Thanks.

**Comment by kevinushey** [comment]
I believe this is fixed in the latest v1.3 patch release (1.3.1056). https://rstudio.com/products/rstudio/download/#download

If you're still seeing issues with that build of RStudio, please let us know and we'll re-open this issue. Thanks!

**Comment by chrisk-edu** [comment]
I am using 1.3.1056 and the behavior persists. setwd fails on a network share (using mapped drive, not UNC), but it works if double-clicking on file (as described before on lwatanabe variation).

**Comment by kevinushey** [comment]
Thanks. Are you able to confirm whether you see the same behavior with R launched within a Windows command prompt, or within RGui (R's bundled R GUI)? Or is this specific to RStudio?

**Comment by williamlief** [comment]
This error is persisting for me in Rstudio 1.3.1056. I can confirm that the problem is only in RStudio. 

When I run `choose.dir()` in the R Gui (r version 4.0.2) I do see the network drives. When I run `choose.dir()` in RStudio I do not see my network drives. If  I follow @lwatanabe post from june 17 by and directly open an r script or project that is on the network drive, then `choose.dir` in RStudio will show the network drives.

**Comment by R-Hannibal** [comment]
Re-opening this ticket based on the feedback from @chrisk-edu & @williamlief that it is still occurring in `1.3.1056.`

@kevinushey - milestone on this ticket is 1.4 and I'm not seeing a PR where it might have been fixed, am I missing something?

**Comment by kevinushey** [comment]
I think I might've erroneously concluded that a previous fix for the handling of UNC paths would have fixed this issue as well. See: https://github.com/rstudio/rstudio/pull/6598

**Comment by mikebessuille** [comment]
Leaving in 1.4 for now

**Comment by williamlief** [comment]
Hi, 
I wanted to follow up with some additional details. It appears that the problem actually happens when I am running as an administrator. R Gui is impacted in the same way. When I open either application with 'run as administrator' they are unable to find the network drives.

I am running Windows 10 (version 18363.959). Happy to provide more details/info if that is helpful. I've had this issue on R 4.0.0 and 4.0.2 and RStudio 1.3.1056

**Comment by kevinushey** [comment]
Thanks! This is useful.

If the error reproduces in RGui, that likely implies the issue lies in R's own code, rather than something specific to RStudio. This limits our ability to solve the problem on the RStudio side, but if we're able to learn enough about what's going on we could attempt to supply a bug report + patch to the R Core team.

**Comment by andzandz11** [comment]
> 
> 
> Thanks! This is useful.
> 

No, this is not related to this bug.

If you run as another user your environmental variables change including your mapped drives. That is intended behaviour under Windows and affects all programs.

**Comment by williamlief** [comment]
> > Thanks! This is useful.
> 
> No, this is not related to this bug.
> 
> If you run as another user your environmental variables change including your mapped drives. That is intended behaviour under Windows and affects all programs.

Thank you! For reference to others who find this thread in the future, running `net use` in the windows cmd prompt will show the mapped network drives. Comparing the output when running `cmd` as an administrator to without will show the differences in what drives can be accessed.

**Comment by kevinushey** [comment]
@andzandz11: The thing that confuses me most about this bug is that, in theory, there shouldn't be anything RStudio-related here -- R's `setwd()` invokes only R's own code, and RStudio's own integration does (or, should not) make a difference here.

This is the R implementation for `setwd()`:

https://github.com/wch/r-source/blob/f21fdfc8b6db5646326c52bcdcbd52ad6357716b/src/main/util.c#L776-L780

If you're able, it might be worth trying to compile a helper C / C++ program that tries to accomplish the same task, with more diagnostics. For example:

```cpp
#include <Rcpp.h>
using namespace Rcpp;

#include <direct.h>

// [[Rcpp::export]]
void cd(std::string path) {
  int status = _chdir(path.c_str());
  if (status != 0) {
    const char* fmt = "error changing working directory [errno=%i]\n";
    Rprintf(fmt, errno);
  }
}

/*** R
cd("Y:/")
*/
```

I am not sure if there are any environment variables (or other application / system state) that could be influence the success or failure here, though.

**Comment by williamlief** [comment]
TLDR I don't think there is any difference between R and RStudio. The problem I was seeing was coming from windows user environmental variables per @andzandz11 .

I had set RStudio to always open as an administrator when I opened it from the task bar, but NOT when I opened a project directly. This led me to believe that opening a project that existed on the network drive was somehow allowing RStudio to see the network drive. However, the actual problem was that I was running RStudio by default with the admin account (and the admin account doesn't have access to the network drive) but when I open a project directly RStudio launches only with regular user credentials and not as an admin. 

When I ran R gui, initially I did not run it as an administrator and it could access the network drive fine. 

I have checked and when I run R or RStudio as an Administrator they cannot see the network drive. When I run either as a regular user they both can access the network drive. Running `net use` in windows cmd with/without administrator level shows that windows cannot/can access the mapped network drives. You can use `net use` to give the admin account access to the network drive.

**Comment by Danny-dK** [comment]
Not sure if this still helps (but the issue is still open). This is not a bug in Windows, this is intended behaviour in Windows. As mentioned by @williamlief, running Rgui or Rstudio in admin mode runs a different set of environment variables / runs in a different context. A quick search on the web "why cmd prompt not show network drives in admin mode" yields many links amongst which:

- https://docs.microsoft.com/en-us/troubleshoot/windows-client/networking/mapped-drives-not-available-from-elevated-command
- https://superuser.com/questions/690287/command-prompt-in-administrator-mode-doesnt-see-mapped-drives
- https://superuser.com/questions/199387/elevated-command-line-prompt-cant-access-shared-drives
- https://superuser.com/questions/495370/why-isnt-a-mapped-drive-available-under-an-elevated-cmd-prompt-but-is-under-a-r

This happens with most windows software that have an elevated start (the search indicates a cmd prompt as this was the easiest search to get to the link, but applies to R and Rstudio).

Some users in this rstudio github issue report it from a university setting. If you have not downloaded and installed R and Rstudio yourself from the associated web pages (because you for example have not received admin or installation privileges from you university), you're most likely getting your R and Rstudio software packaged from the university. Most universities have a separate environment in which they offer packaged software to employees that can be installed, for example from something like a university "software center". These packaged software can be installed from a university link without admin rights required and have various setting pre-registered. If this is the case for you, you need to contact your own IT departement to make them adjust these setting. Most likely they have set admin rights to the software to be able to install packages into the C-drive library. In my university we had many issues with the packaged university R and Rstudio versions, but were solved (by setting appropriate group policies and environment variables).

Hope this info helped.

**Comment by gwd666** [comment]
@twang96 
> > My solution...save a script on the network drive and open the script directly (double-click). It opened with no problem. Now, RStudio can find the drive, even after restarting my machine.
> 
> This works. Try it.

What you are witnessing as "working" is )most of the time) "just" a correct connecting to a mapped network drive getting established;  
meaning (ie if you witthness that sort of "bug" even when using a RStudio version that has all the necessary patches for the "real" bug that started this issue thread included) it will usually just be your network drive not being correctly connected to your machine - we also have that regularly (especially if there was/is a VPN involved during start-up and network drive connection, which was quite common in HO times). 
So in that case when you save a script on your network share you are usually (mainly) fixing that network connection error, meaning for your current session you are clicking through to some network folder and that will usually suffice to make that connection share drive "live" or cleanly connect and RStudio and/or  R  (usually after a restart) will also pick up the network share correctly again.   
Means, most of the time you can save yourself the extra effort of saving files, if your R/RStudio session is "unaware" of the network share, just close it click on the shared drive letter in your Windows Explorer (to fix that incomplete drive connection) and then open R/RStudio again.
